pkg/lib/cockpit-components-table.jsx

```diff
@@ -63,325 +56,190 @@ import './cockpit-components-table.scss';
-                return emptyState;
-
-            const emptyStateCell = (
-                [{
-                    props: { colSpan: this.props.columns.length },
-                    title: emptyState
-                }]
-            );
-
-            if (isTableBasic)
-                tableProps.rows = [{ columns: emptyStateCell }];
-            else
-                tableProps.rows = [{ cells: emptyStateCell }];
-        }
-
-        if (this.props.style)
-            tableProps.style = this.props.style;
-
-        if (!isTableBasic) {
-            return (
-                <Table {...tableProps}>
-                    {this.props.showHeader && <TableHeader />}
-                    <TableBody {...tableBodyProps} />
-                </Table>
-            );
-        } else {
-            return <ComposableTableBasic showHeader={this.props.showHeader} {...tableProps} />;
-        }
-    }
-}
-ListingTable.defaultProps = {
-    caption: '',
-    emptyCaption: '',
-    isEmptyStateInTable: false,
-    columns: [],
-    rows: [],
-    actions: [],
-    showHeader: true,
-};
-ListingTable.propTypes = {
-    id: PropTypes.string,
-    caption: PropTypes.string,
-    emptyCaption: PropTypes.node,
-    emptyCaptionDetail: PropTypes.node,
-    isEmptyStateInTable: PropTypes.bool,
-    columns: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.object, PropTypes.string])),
-    rows: PropTypes.arrayOf(PropTypes.shape({ props: PropTypes.object })),
-    actions: PropTypes.node,
-    variant: PropTypes.string,
-    showHeader: PropTypes.bool,
-    style: PropTypes.object,
-    afterToggle: PropTypes.func,
-};
+                        return (
+                            <Td key={key || `row_${rowKey}_cell_${dataLabel}`} dataLabel={dataLabel} {...cellProps}>
+                                {typeof cell == 'object' ? cell.title : cell}
+                            </Td>
+                        );
+                    })}
+                </Tr>
+                {row.expandedContent && <Tr id={"expanded-content" + rowIndex} isExpanded={isExpanded}>
+                    <Td noPadding={row.hasPadding !== true} colSpan={row.columns.length + 1 + (onSelect ? 1 : 0)}>
+                        <ExpandableRowContent>{row.expandedContent}</ExpandableRowContent>
+                    </Td>
+                </Tr>}
+            </React.Fragment>
+        );
+
+        if (row.expandedContent)
+            return <Tbody key={rowKey} isExpanded={isExpanded}>{rowPair}</Tbody>;
+        else
+            return rowPair;
+    });
 
-const ComposableTableBasic = ({
-    cells,
-    header,
-    rows,
-    showHeader,
-    ...props
-}) => {
     return (
         <>
             {header}
-            <TableComposable {...props}>
+            <TableComposable {...extraProps} {...tableProps}>
                 {showHeader && <Thead>
                     <Tr>
+                        {isExpandable && <Th />}
+                        {onSelect && <Th />}
```

Comment:
Are these empty table headers just "filling" to get the expected number of columns? But they are class-less, how do they actually render the expander sign/select checkbox?

Reply:
When the table has checkboxes or expanders on rows, (first cell of row occupied) the header columns need to move 1 or 2 positions (when both expander and checkbox exist) right accordingly. I checked how they do it in the PF4 examples, and they just render empty th as we do here.

Similarly, if one row is expandable and the next row is not, (we do that in the storage page in content views), we need to move the first column of the non - expandable row one position right, so that it aligns with the first expandable row.

Hope this makes sense.

Reply:
Oh right, these are the expanded rows. Thanks!